modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/statistic/SqlStatisticManagerImpl.java

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.statistic;
+
+import static org.apache.ignite.internal.event.EventListener.fromConsumer;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.catalog.CatalogManager;
+import org.apache.ignite.internal.catalog.events.CatalogEvent;
+import org.apache.ignite.internal.catalog.events.DropTableEventParameters;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.table.TableViewInternal;
+import org.apache.ignite.internal.table.distributed.TableManager;
+import org.apache.ignite.internal.util.FastTimestamps;
+import org.apache.ignite.internal.util.Pair;
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * Statistic manager. Provide and manage update of statistics for SQL.
+ */
+public class SqlStatisticManagerImpl implements SqlStatisticManager {
+    private static final IgniteLogger LOG = Loggers.forClass(SqlStatisticManagerImpl.class);
+    private static final long DEFAULT_TABLE_SIZE = 1_000_000;
+    private static final long MINIMUM_TABLE_SIZE = 1_000L;
+    private static final Pair<Long, Long> DEFAULT_VALUE = new Pair<>(DEFAULT_TABLE_SIZE, 0L);
+
+    private final TableManager tableManager;
+
+    private final ConcurrentMap<Integer, Pair<Long, Long>> tableSizeMap = new ConcurrentHashMap<>();
+
+    private volatile long thresholdTimeToPostponeUpdateMs = TimeUnit.MINUTES.toMillis(1);
+
+    /** Constructor. */
+    public SqlStatisticManagerImpl(TableManager tableManager, CatalogManager catalogManager) {
+        this.tableManager = tableManager;
+
+        catalogManager.listen(CatalogEvent.TABLE_DROP, fromConsumer(event -> {
+            int tableId = ((DropTableEventParameters) event).tableId();
+            onTableDrop(tableId);
+        }));
+    }
+
+    /**
+     * Returns approximate number of rows in table by their id. Returns the previous known value or
+     * {@value SqlStatisticManagerImpl#DEFAULT_TABLE_SIZE} as default value. Can start process to update asked statistics in background to
+     * have updated values for future requests.
+     */
+    @Override
+    public long tableSize(int tableId) {
+        updateTableSizeStatistics(tableId);
+        long ts = tableSizeMap.getOrDefault(tableId, DEFAULT_VALUE).getFirst();
+
+        return Math.max(ts, MINIMUM_TABLE_SIZE);
+    }
+
+    /** Update table size statistic in the background if it required. */
+    private void updateTableSizeStatistics(int tableId) {
+        tableSizeMap.putIfAbsent(tableId, DEFAULT_VALUE);
```

Comment:
does it make sense to acquire TableViewInterna first, and, if it's available, proceed with updating?

Reply:
agree, fixed